components/proxy/src/main/kotlin/com/hotels/styx/services/OriginsAdminHandler.kt

```diff
@@ -0,0 +1,172 @@
+        return if (origin != null) {
+            textValueResponse(stateTagValue(origin.tags) ?: "")
+        } else {
+            errorResponse(NOT_FOUND, "No origin found for ID $objectId")
+        }
+    }
+
+    private fun putState(objectId: String, request: HttpRequest) : HttpResponse =
+            try {
+                val body = request.bodyAs(UTF_8)
+                val value = MAPPER.readValue(body, String::class.java)
+                try {
+                    when (value) {
+                        STATE_ACTIVE -> activate(objectId)
+                        STATE_INACTIVE -> close(objectId)
+                        else -> errorResponse(BAD_REQUEST, "Unrecognized target state: $value")
+                    }
+                } catch (e: HttpStatusException) {
+                    errorResponse(e.status, e.message)
+                }
+            } catch (t: Throwable) {
+                errorResponse(BAD_REQUEST, "Error handling state change request: ${t.localizedMessage}")
+            }
+
+    private fun activate(objectId: String) : HttpResponse {
+        var newState = ""
+        routeDb.compute(objectId) { origin ->
+            if (!isValidOrigin(origin)) {
+                throw HttpStatusException(NOT_FOUND, "No origin found for ID $objectId")
+            }
+            newState = when(stateTagValue(origin!!.tags)) {
+                STATE_INACTIVE, STATE_ACTIVE -> STATE_ACTIVE
+                STATE_UNREACHABLE -> STATE_UNREACHABLE
+                else -> STATE_ACTIVE
+            }
+            updateStateTag(origin, newState)
+        }
+        return HttpResponse.response(OK)
+                .body(MAPPER.writeValueAsString(newState), UTF_8)
+                .build()
+    }
+
+    private fun close(objectId: String) : HttpResponse {
+        routeDb.compute(objectId) { origin ->
+            if (!isValidOrigin(origin)) {
+                throw HttpStatusException(NOT_FOUND, "No origin found for ID $objectId")
+            }
+            updateStateTag(origin!!, STATE_INACTIVE, true)
+        }
+        return HttpResponse.response(OK)
+                .body(MAPPER.writeValueAsString(STATE_INACTIVE), UTF_8)
+                .build()
+    }
+
+    private fun updateStateTag(origin: RoutingObjectRecord, newValue: String, clearHealthcheck: Boolean = false) : RoutingObjectRecord {
+        val oldTags = origin.tags
+        val newTags = oldTags
+                .filterNot{ clearHealthcheck && isHealthCheckTag(it) }
+                .filterNot(::isStateTag)
+                .plus(stateTag(newValue)).toSet()
+        return if (oldTags != newTags) {
+            origin.copy(tags = newTags)
+        } else {
+            origin
+        }
+    }
+
+    private fun errorResponse(status: HttpResponseStatus, message: String) =
+            HttpResponse.response(status)
+                    .disableCaching()
+                    .addHeader(HttpHeaderNames.CONTENT_TYPE, JSON_UTF_8.toString())
```

Comment:
As per coding style we would static inline constants like `CONTENT_TYPE` and `response`.